spec/models/queries/projects/project_queries/scopes/allowed_to_spec.rb

```diff
@@ -0,0 +1,127 @@
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+require "spec_helper"
+
+RSpec.describe Queries::Projects::ProjectQuery, "#allowed to" do # rubocop:disable RSpec/FilePath,RSpec/SpecFilePathFormat
+  shared_let(:user) { create(:user) }
+  shared_let(:other_user) { create(:user) }
+
+  let(:checked_user) { user }
+  let(:permission) { :view_project_query }
+
+  subject { described_class.allowed_to(checked_user, permission) }
+
+  shared_let(:view_role) { create(:view_project_query_role) }
+  shared_let(:edit_role) { create(:edit_project_query_role) }
+
+  shared_let(:owned_query) { create(:project_query, user:) }
+  shared_let(:owned_public_query) { create(:project_query, user:, public: true) }
+  shared_let(:public_other_query) { create(:project_query, user: other_user, public: true) }
+  shared_let(:private_other_query) { create(:project_query, user: other_user) }
+  shared_let(:private_other_query_with_view) do
+    create(:project_query, user: other_user, members: [
+             create(:member, user:, roles: [view_role])
+           ])
+  end
+  shared_let(:private_other_query_with_edit) do
+    create(:project_query, user: other_user, members: [
+             create(:member, user:, roles: [edit_role])
+           ])
+  end
+
+  context "when the user is locked" do
+    let(:checked_user) { create(:locked_user) }
+
+    it { is_expected.to be_empty }
+  end
+
+  context "when no permission is checked" do
+    let(:permission) { nil }
+
+    it { is_expected.to be_empty }
+  end
+
+  context "when the user is anonymous" do
+    let(:checked_user) { create(:anonymous) }
+
+    it { is_expected.to be_empty }
+  end
+
+  context "for the view permission" do
+    let(:permission) { :view_project_query }
+
+    it do
+      expect(subject).to contain_exactly(
+        # public queries
+        public_other_query,
+        owned_public_query,
+        # user owned queries
+        owned_query,
+        # view membership queries
+        private_other_query_with_view,
+        private_other_query_with_edit
+      )
+    end
+  end
+
+  context "for the edit permission" do
+    let(:permission) { :edit_project_query }
+
+    context "when the user can manage global queries" do
+      before do
+        mock_permissions_for(user) do |mock|
+          mock.allow_globally(:manage_public_project_queries)
+        end
+      end
+
+      it do
+        expect(subject).to contain_exactly(
+          # public queries
+          public_other_query,
+          owned_public_query,
+          # user owned queries
+          owned_query,
+          # view membership queries
+          private_other_query_with_edit
+        )
+      end
+    end
+
+    context "when the user cannot manage global queries" do
+      it do
+        expect(subject).to contain_exactly(
+          # user owned queries
+          owned_query,
```

Comment:
Shouldn't we also have `owned_public_query` here?

Reply:
Here we are checking for the edit permission. Public queries are only editable when you have the `manage_public_project_queries` permission. Should I make this a bit more obvious by adding a comment?

Reply:
I think the description is fine, I just find it weird that a user who created a public query is not able to edit it unless it has the `manage_public_project_queries`. I feel like users should be able to edit their own queries regardless if the query is public or private. But maybe this is more of a product discussion.

Reply:
As soon as the list is made public, the user basically "loses" ownership. As from this moment on it is accessible to everybody, a user could potentially do "some harm" by changing the list that is then used by everybody. So I think this behavior makes sense. This is also how the permission is currently implemented